### PR TITLE
Fix incorrect source gen in edge cases of dependency injection

### PR DESCRIPTION
**Pull request by smoogipoo**
Fixes the test errors in https://github.com/ppy/osu-framework/pull/5543

Tested with o-f and osu.

### CODE REVIEW

File: osu.Framework.SourceGeneration/SyntaxHelpers.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public static InvocationExpressionSyntax GetDependencyInvocation(ITypeSymbol cal
                                                          SyntaxFactory.TypeArgumentList(
                                                              SyntaxFactory.SeparatedList(new[]
                                                              {
-                                                                 SyntaxFactory.ParseTypeName(GetUnderlyingType(requestedType))
+                                                                 SyntaxFactory.ParseTypeName(requestedType.ToDisplayString())
```
Comment (smoogipoo) [inline osu.Framework.SourceGeneration/SyntaxHelpers.cs:75]: The reason this was `GetUnderlyingType` before is because the source gen didn't emit NRT-enabled files at one point in time, and in such contexts `object?` is an invalid expression. But now this is fine since it does emit NRT-enabled files.

File: osu.Framework.SourceGeneration/SyntaxContextReceiver.cs [hunk 2]
```diff
@@ -48,9 +48,9 @@ public void OnVisitSyntaxNode(GeneratorSyntaxContext context)
             }
 
             // Process any [Cached] attributes on any interface on the class excluding base types.
-            foreach (var iFace in symbol.Interfaces)
+            foreach (var iFace in SyntaxHelpers.GetDeclaredInterfacesOnType(symbol))
```
Comment (smoogipoo) [inline osu.Framework.SourceGeneration/SyntaxContextReceiver.cs:51]: The distinction here of "declared interfaces" is actually very important. One particular case osu! side looks like the following (simplified):

```csharp
// In the osu.Game project:

[Cached]
internal interface IOverlayManager { } // Note the use if internal.
public abstract class PlayerTestScene : IOverlayManager { }

// In the ruleset project:

public class TestSceneOsuPlayer : PlayerTestScene { }
```

"Declared" is important because `TestSceneOsuPlayer` cannot cache itself as an `IOverlayManager` due to not having access to that interface. It has to rely on the declarer of that interface (`PlayerTestScene`) to do that work.

**Review by smoogipoo — COMMENTED** [review]
Just explaining a few things, in-case they're not self-explanatory.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]
Tested to fix the failures via a local nuget package.
